docs/flip/1009-optimizer-api.md

```diff
@@ -0,0 +1,529 @@
+    (loss, new_model_state), grads = jax.value_and_grad(
+        loss_from_params, has_aux=True)(params)
+    new_params, new_opt_state = self.optim.update(grads, params, self.opt_state)
+
+    return self.replace(
+        variables=variables.copy(dict(params=new_params, **new_model_state)),
+        opt_state=new_opt_state,
+    ), loss
+```
+
+Users can derive from this dataclass and put any additional state into that
+object:
+
+```python
+@flax.struct.dataclass
+class TrainState(LinenHelper):
+  step: int
+```
+
+This will reduce above training loop to something much simpler:
+
+```python
+@jax.jit
+def train_step(state, inputs, labels):
+
+  def loss_fn(inputs, outputs, labels):
+    del inputs  # unused
+    return xent_loss(outputs, labels)
+
+  state, loss = state.replace(step=state.step + 1).update_with_loss(
+      loss_fn, inputs, labels, mutable=['batch_stats'])
+  return state, loss
+
+opt_state = optim.init(variables['params'])
+state = TrainState(
+  apply_fn=model.apply,
+  variables=variables,
+  optim=optim,
+  opt_state=opt_state,
+  step=0,
+)
+for batch in ds.as_numpy_iterator():
+  state, loss = train_step(state, batch['image'], batch['label'])
+  print(loss)
+```
+
+Remarks:
+
+- It is quite common to have `TrainState`-like objects in Flax that keep all
+  the state (and some functions) and are used for checkpointing.
+- Variables can be inspected and additionally modified in `train_step()` if
+  needed.
+- `LinenHelper` might provide more utility functions with slightly different
+  signatures if there is demand for that.
+
+# Previous API
+[previous API]: #previous-api
+
+## Previous Implementation
+[Previous Implementation]: #previous-implementation
+
+The optimizer itself would be implemented by creating a new class deriving
+from `OpimizerDef`:
+
+```python
+# flax/optim/momentum.py
+
+@flax.struct.dataclass
+class _MomentumHyperParams:
+  learning_rate: jnp.ndarray
+  beta: jnp.ndarray
+
+
+@flax.struct.dataclass
+class _MomentumParamState:
+  momentum: np.ndarray
+
+
+class Momentum(flax.optim.OptimizerDef):
+
+  def __init__(self, learning_rate=None, beta=0.9):
+    super().__init__(
+      _MomentumHyperParams(learning_rate, beta)
+    )
+
+  def init_param_state(self, param):
+    return _MomentumParamState(jnp.zeros_like(param))
+
+  def apply_param_gradient(self, step, hyper_params, param, state, grad):
+    del step
+    assert hyper_params.learning_rate is not None
+    new_momentum = state.momentum * hyper_params.beta + grad
+    new_params = param - hyper_params.learning_rate * new_momentum
+    return new_params, _MomentumParamState(new_momentum)
+```
+
+Remarks:
+
+- Note the relationship between `OptimizerDef` and `Optimizer` : When the
+  function `Optimizer.apply_gradient()` is called from the user code, it calls
+  into `OptimizerDef.apply_gradient()` (among other things) which in turn will
+  call `OptimizerDef.apply_param_gradient()` which is implemented by subclasses
+  of `OptimizerDef`.
```

Comment:
Suggestion: nit 

(when tryin to follow, present & future tense mixing may be a bit confusing)
("which.... which..." -> "which... (...)")

```suggestion
- Note the relationship between `OptimizerDef` and `Optimizer`: When the
  function `Optimizer.apply_gradient()` is called from the user code, it calls
  into `OptimizerDef.apply_gradient()` (among other things) which in turn
  calls `OptimizerDef.apply_param_gradient()` (implemented by subclasses
  of `OptimizerDef`).
```

Reply:
Updated.